### PR TITLE
[SYCL][CUDA][HIP] Device info query for maximum registers per block and targetted exception on out-of-registers for CUDA.

### PR DESCRIPTION
**Pull request by GeorgeWeb**
This PR introduces the following changes:
- Adds new device info trait to query max number of registers per wg/block.
  - Includes e2e test for the `max_registers_per_work_group` query. (cuda and hip only)
- Propagates and handles error on out-of-registers on detection of exceeded register use, which gives better clue to why the kernel can't launch, an improvement over an obscure out of launch resources Cuda error code.
  - Includes e2e test with a kernel that uses a lot of registers to exceed max registers per block, which tests the error case detection and corresponding sycl rt exception. (cuda only)

This is related to the following issue: https://github.com/intel/llvm/issues/8414, as well as another in Blender.

---

There will be a follow-up PR to allow setting the CUjit option for limiting number of registers used per thread, but this PR is to catch and better communicate the error, as well as allowing to utilise `max_registers_per_work_group` when calculating the local/wg sizes for `nd_range`.

(Note: Ideally, I would split this in two separate PRs, the second dependent on the first.)

---

Signed-off-by: GeorgeWeb (Georgi Mirazchiyski) [georgi.mirazchiyski@codeplay.com](mailto:georgi.mirazchiyski@codeplay.com)

### CODE REVIEW

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 1]
```diff
@@ -3159,10 +3195,18 @@ pi_result cuda_piEnqueueKernelLaunch(
...
+        size_t kernelLocalWorkGroupSize = 0;
         for (size_t dim = 0; dim < work_dim; dim++) {
           auto err = isValid(dim);
           if (err != PI_SUCCESS)
             return err;
+          // If no error then sum the total local work size per dim.
+          kernelLocalWorkGroupSize += local_work_size[dim];
+        }
+
+        if (hasExceededMaxRegistersPerBlock(command_queue->device_, kernel,
```
[the remark this answers is not in the document]
Comment (jchlanda) [reply sycl/plugins/cuda/pi_cuda.cpp:3213]: Sounds good, feel free to do it in a follow up PR, if you're keen to get this one in first.

File: sycl/include/sycl/detail/pi.h [hunk 2]
```diff
@@ -347,6 +347,8 @@ typedef enum {
   PI_EXT_CODEPLAY_DEVICE_INFO_SUPPORTS_FUSION = 0x20005,
   PI_EXT_DEVICE_INFO_ATOMIC_FENCE_ORDER_CAPABILITIES = 0x20006,
   PI_EXT_DEVICE_INFO_ATOMIC_FENCE_SCOPE_CAPABILITIES = 0x20007,
+  // The number of max registers per block (device specific)
+  PI_EXT_CODEPLAY_DEVICE_INFO_MAX_REGISTERS_PER_WORK_GROUP = 0x20008
```
Comment (steffenlarsen) [inline sycl/include/sycl/detail/pi.h:351]: This is an addition to the PI interface and as such you need to increment `_PI_H_VERSION_MINOR` and add a version note above it.

File: sycl/include/sycl/info/ext_codeplay_device_traits.def [hunk 3]
```diff
@@ -3,6 +3,9 @@
 #define __SYCL_PARAM_TRAITS_TEMPLATE_SPEC __SYCL_PARAM_TRAITS_SPEC
 #endif
 __SYCL_PARAM_TRAITS_SPEC(ext::codeplay::experimental,device, supports_fusion, bool, PI_EXT_CODEPLAY_DEVICE_INFO_SUPPORTS_FUSION)
+__SYCL_PARAM_TRAITS_SPEC(
+    ext::codeplay::experimental, device, max_registers_per_work_group, uint32_t,
```
Comment (steffenlarsen) [inline sycl/include/sycl/info/ext_codeplay_device_traits.def:7]: Where is this extension documented?

File: sycl/source/detail/error_handling/error_handling.cpp [hunk 4]
```diff
@@ -230,6 +233,47 @@ void handleInvalidWorkGroupSize(const device_impl &DeviceImpl, pi_kernel Kernel,
...
+
+      uint32_t NumRegisters = 0;
+      Plugin.call<PiApiKind::piKernelGetGroupInfo>(
+          Kernel, Device, PI_KERNEL_GROUP_INFO_NUM_REGS, sizeof(NumRegisters),
+          &NumRegisters, nullptr);
+
+      uint32_t MaxRegistersPerBlock = 0;
+      Plugin.call<PiApiKind::piDeviceGetInfo>(
+          Device, PI_EXT_CODEPLAY_DEVICE_INFO_MAX_REGISTERS_PER_WORK_GROUP,
+          sizeof(MaxRegistersPerBlock), &MaxRegistersPerBlock, nullptr);
```
Comment (steffenlarsen) [inline sycl/source/detail/error_handling/error_handling.cpp:256]: We might as well use the new info query here.
```suggestion
      uint32_t MaxRegistersPerBlock =
          DeviceImpl.get_info<ext::codeplay::experimental::info::device::
                                  max_registers_per_work_group>();
```

File: sycl/include/sycl/detail/pi.h (hunk 2 above)
Comment (GeorgeWeb) [reply sycl/include/sycl/detail/pi.h]: Added. :+1:

File: sycl/include/sycl/info/ext_codeplay_device_traits.def (hunk 3 above)
Comment (GeorgeWeb) [reply sycl/include/sycl/info/ext_codeplay_device_traits.def:7]: I've now added `sycl/doc/extensions/experimental/sycl_ext_codeplay_max_registers_per_work_group_query.md`. I have seen another extension for `num_regs` having just an inline documentation where it was added in the PI header, but thought I'd add a full doc for `sycl_ext_codeplay_max_registers_per_work_group` with a MACRO.
Does that sound okay to you? And if so, and I am missing something, please let me know. Thank you! :)

File: sycl/doc/extensions/experimental/sycl_ext_codeplay_max_registers_per_work_group_query.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+Executed using the CUDA back-end on NVIDIA.
+
+```
+NVIDIA ...
+Max registers per work-group: 65536
+```
+
+- See: [CUDA Toolkit Documentation](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#compute-capabilities)
+
+## Implementation
```
Comment (steffenlarsen) [inline sycl/doc/extensions/experimental/sycl_ext_codeplay_max_registers_per_work_group_query.md:51]: I would prefer if we just drop this section.

File: sycl/include/sycl/info/ext_codeplay_device_traits.def (hunk 3 above)
Comment (steffenlarsen) [reply sycl/include/sycl/info/ext_codeplay_device_traits.def:7]: Thank you! Yes, I am okay with that and I think I see the one you are referring to. That one might have slipped through the cracks, but the general rule is that public interfaces like that should either be specification or extension interfaces.

@AerialMantis - On a related note, are there any particular rhyme or reason to when we use `codeplay` in the extension features vs. `intel` and/or `oneapi`?

File: sycl/doc/extensions/experimental/sycl_ext_codeplay_max_registers_per_work_group_query.md (hunk 5 above)
Comment (steffenlarsen) [reply sycl/doc/extensions/experimental/sycl_ext_codeplay_max_registers_per_work_group_query.md]: That includes its sub-section(s). 😉

Comment (GeorgeWeb) [reply sycl/doc/extensions/experimental/sycl_ext_codeplay_max_registers_per_work_group_query.md]: Ah yeah, excuse me. Never good to rush work on Fridays.😅 Cheers for double checking that.

File: sycl/test-e2e/Basic/max_registers_per_work_group_query.cpp [hunk 6]
```diff
@@ -0,0 +1,24 @@
+// REQUIRES: cuda || hip
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
```
Comment (steffenlarsen) [inline sycl/test-e2e/Basic/max_registers_per_work_group_query.cpp:3]: Please change to using the new testing format, i.e. something like:

```suggestion
// RUN: %{build} -o %t.out
// RUN: %{run} %t.out
```

**Review by steffenlarsen — COMMENTED** [review]
Overall I think it looks good, but we need the tests changed to the new format.

File: sycl/doc/extensions/experimental/sycl_ext_codeplay_max_registers_per_work_group_query.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+## Notice
+
+This document describes an **experimental** API that applications can use to try
+out a new feature. Future versions of this API may change in ways that are
+incompatible with this experimental version.
+
+
+## Introduction
+
+This extension adds functionally a new device information descriptor. It provides the ability to query a device for the maximum number of registers available per work-group.
```
Comment (steffenlarsen) [inline sycl/doc/extensions/experimental/sycl_ext_codeplay_max_registers_per_work_group_query.md:12]: Nit; I think this reads a little better.
```suggestion
This extension adds a new device information descriptor that provides the ability to query a device for the maximum number of registers available per work-group.
```

**Review by steffenlarsen — APPROVED** [review]
Small nit, otherwise LGTM!

File: sycl/doc/extensions/experimental/sycl_ext_codeplay_max_registers_per_work_group_query.md (hunk 7 above)
Comment (GeorgeWeb) [reply sycl/doc/extensions/experimental/sycl_ext_codeplay_max_registers_per_work_group_query.md:12]: Does sound much clearer. Appreciated!

**Review by steffenlarsen — APPROVED** [review]

**Pull request merged**